### PR TITLE
VSI initial

### PR DESCRIPTION
**Pull request by jssblck**
Initial VSI functionality.

Missing features:
- Filtering
- Unarchiving

These will be added in a separate PR.

# Testing plan

Automated tests are run in CI: https://github.com/fossas/foundation-libs/pull/4/files#diff-c663abea89dc360ec0b9086f89d4940d610795fa2d0f8ce351bd367da79dc3db

Windows does not currently work for the integration test (this is a known issue: see https://fossa.atlassian.net/browse/ANE-580). This is also due to be fixed in another PR (basically, once I get the chance to come back to this).

### CODE REVIEW

File: fingerprint/src/fingerprint.rs [hunk 1]
```diff
@@ -26,13 +26,23 @@ pub fn comment_stripped<R: Read>(
...
     // Chain the part of the stream already read to evaluate binary along with the rest of the stream.
     let mut stream = Cursor::new(read).chain(stream);
-    if is_binary {
-        Ok(None)
-    } else {
-        Some(hash_text_stripped(&mut stream)).transpose()
+    match hash_text_stripped(&mut stream) {
+        Ok(fp) => Ok(Some(fp)),
+        Err(err) => {
+            // The `io::Error` type is opaque 😭
```
Comment (meghfossa) [inline fingerprint/src/fingerprint.rs]: [no action] Github can't render emoji in diffs :( 
CleanShot 2022-10-13 at 10 41 58@2x

File: vsi/src/api/fossa.rs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+
+        let url = self.endpoint.join("scans")?;
+        let req = self
+            .client
+            .post(url.clone())
+            .bearer_auth(&self.api_key)
+            .json(&req_body);
+
+        let res_body = run_req!(download, req, url, req_body);
+        // let res = req.send().await.context("send request")?;
```
Comment (meghfossa) [inline vsi/src/api/fossa.rs]: [nit] dead comments?

File: vsi/src/api/fossa.rs [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+/// Parse JSON text into a type wrapped with a context displaying the input on error.
+macro_rules! parse {
+    ($input:expr) => {
+        from_str(&$input).wrap_err_with(|| format!("parse: {}", $input))
+    };
+}
+
+/// Run a request, returning its body as a string on success.
+macro_rules! run_req {
```
Comment (meghfossa) [inline vsi/src/api/fossa.rs:31]: [no action] :star:

File: fingerprint/Cargo.toml [hunk 4]
```diff
@@ -1,18 +1,19 @@
 [package]
 name = "fingerprint"
-version = "0.1.0"
+version = "1.0.0"
 edition = "2021"
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+[features]
+default = []
```
Comment (skilly-lily) [inline fingerprint/Cargo.toml]: [nit]: Do we need this?  I think it can removed.

File: vsi/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+async-trait = "0.1.57"
+atty = "0.2.14"
+cancel = "0.1.0"
+clap = { version = "4.0.4", features = ["derive", "cargo", "env"] }
+defer-lite = "1.0.0"
+derive_more = "0.99.17"
+getset = "0.1.2"
+lazy_static = "1.4.0"
+log = "0.4.17"
+mockall = "0.11.2"
```
Comment (skilly-lily) [inline vsi/Cargo.toml]: [optional nit]: can this be a dev-dependency?  We can implement it with `#[cfg_attr(test, mockall)]`

File: vsi/src/api.rs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    async fn download_forensics(&self, id: &scan::Id) -> Result<HashSet<Locator>>;
+}
+
+/// Blanket implement anything that implements a thread safe client as a scan sink.
+#[async_trait]
+impl<T: Client + Sync> scan::Sink for T {
+    type Id = scan::Id;
+
+    async fn append_scan(&self, id: &Self::Id, artifacts: Vec<scan::Artifact>) -> Result<()> {
+        self.append_scan(id, artifacts).await
```
Comment (skilly-lily) [inline vsi/src/api.rs]: [nit]: We should adjust the names to be slightly different, this is a bit confusing to readers.

File: vsi/src/forensics.rs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    Failed,
+    /// Forensic analysis is in some other status, which can be displayed to the user.
+    /// This status indicates that it is in process.
+    Informational(String),
+}
+
+impl Status {
+    /// Safely parse the string.
+    pub fn parse(input: String) -> Self {
+        match input.as_str() {
```
Comment (meghfossa) [inline vsi/src/forensics.rs:27]: [optional] are these matches case-sensitive? if not, prefer if cast them upper or lower or do case-insensitive match

File: vsi/src/api/fossa.rs (hunk 2 above)
Comment (meghfossa) [reply vsi/src/api/fossa.rs]: there are many cases of this in this file

**Review by meghfossa — COMMENTED** [review]
I'm only 1/6'th of the way through - but leaving some comments for now

File: fingerprint/src/fingerprint.rs (hunk 1 above)
Comment (jssblck) [reply fingerprint/src/fingerprint.rs]: Yeah, removed :(

File: vsi/src/forensics.rs (hunk 7 above)
Comment (jssblck) [reply vsi/src/forensics.rs:27]: Hmm, they're just an enum server side. So not really case insensitive, but why do the extra work of rendering them lower/upper?

Comment (skilly-lily) [reply vsi/src/forensics.rs:27]: If the server can make the contractual guarantee about casing, then we don't need to do case-insensitive matching.  In cases where the server might change without warning and we'd have to be defensive against that, then case-insensitive matching provides a relatively cheap defense mechanism.

If sherlock guarantees casing, then I don't think this is a big deal.  if it's loose about casing, then we should either start a contract, or match defensively here.

Comment (skilly-lily) [reply vsi/src/forensics.rs:27]: To clarify: a "contract" might just be a server-side comment that says "these must be upper-case, as clients expect uppercase values" or something similar.

**Review by skilly-lily — APPROVED** [review]
Few comments, nothing blocking

File: vsi/src/forensics.rs (hunk 7 above)
Comment (jssblck) [reply vsi/src/forensics.rs:27]: These are treated as an enum, and tons of server internals depend on them and are case sensitive: https://github.com/fossas/basis/blob/be26d1cb524f0ebb0128d34eaccc5d6903604a3d/sherlock/internal/models/forensics_pipeline.go#L6-L17

Comment (skilly-lily) [reply vsi/src/forensics.rs:27]: Good enough for me, I think we can call that a contract.

**Pull request merged**